admin/public/js/components/charts.jsx

```diff
@@ -0,0 +1,102 @@
+      
+      return {
+        maxVal: parseFloat(this.props.ydomain[0]),
+        minVal: parseFloat(this.props.ydomain[1]),
+        data: data,
+        lineData: ""
+      };
+
+    },
+
+    getDefaultProps: function() {
+      return {
+        title: "Chart", 
+        ydomain: [0, 100],
+        xdomain: [1, 300], 
+        interpolation: "basis",
+        width: 400,
+        y_width: 50,
+        height: 90,
+        value: 0,
+        animationRate: 40
+      }
+    },
+
+    render: function() {
+
+      return (
+        <table cellPadding="0" cellSpacing="0" className="graph">
+        <tbody>
+          <tr>
+            <td className="header title">
+              <span className="label label-default">{this.props.title}</span>
+            </td>
+          </tr>
+          <tr>
+            <td>
+              <svg width={this.props.width} height={this.props.height}>
+                <g>
+                  <defs><clipPath id="clip"><rect x={this.props.y_width} width={this.props.width-this.props.y_width} height={this.props.height}></rect></clipPath></defs>
+                  <g clipPath="url(#clip)"><path ref="path" className="line" d={this.state.lineData} transform=""></path></g>
+                  <g className="y_axis" ref="y_axis" transform={ "translate("+this.props.y_width+",0)" }></g>
+                </g>
+              </svg>
+            </td>
+            <td className="values">
+              <div><span className="label label-primary">Current</span><span className="label label-primary">{this.props.value}</span></div>
+              <div><span className="label label-info">Min</span><span className="label label-info">{this.state.minVal}</span></div>
+              <div><span className="label label-info">Max</span><span className="label label-info">{this.state.maxVal}</span></div>
+            </td>
+
+          </tr>
+        </tbody>
+        </table>
+      );
+    },
+
+    updateGraphData: function() {
+      var newPoint = this.props.value;
+      var newData = this.state.data.slice(1).concat([newPoint]);
+      var newState = { data: newData, lineData: this.line(newData), source:"internal"};
+      
+      if (newPoint > this.state.maxVal) {
+        newState.maxVal = newPoint;
+      } else if (newPoint < this.state.minVal) {
+        newState.minVal = newPoint;
+      }
+      this.setState(newState);
+    },
+
+    componentDidMount: function() {
+      d3.select(this.refs.y_axis).call(d3.svg.axis().scale(this.y).ticks(5).orient("left"));
+    },
+
+    componentDidUpdate: function () {
+      var that = this;
+      // redraw the line, and then slide it to the left
+      d3.select(that.refs.path)
+        .transition()
+        .duration(that.props.animationRate)
+        .ease("linear")
+        .attr("transform", "translate(" + that.x(0) + ")")
+        .each("end", function() { that.updateGraphData(); });
+    }
```

Comment:
Overall this component looks great. I think this function however causes an infinite loop; a re-render calls `componentDidUpdate`, which calls `updateGraphData`, which calls `setState`, which causes another re-render. I'm not 100% if that's quite what you want or not but here's another thought.

Since you want the graph to update on a regular interval, you might consider completely breaking out of the React update cycle and implementing it completely yourself. You could do this by the code in this function to a separate function and invoke it on a timer; super-pseudocode, but something like

```javascript
shouldComponentUpdate() {
  // tell react it should *never* re-render this component
  // automatically when state or props change
  return false;
}

componentDidMount() {
  updateGraph();
}

updateGraph() {
  d3.select(this.refs.path)
    .transition()
    .duration(this.props.animationRate)
    .ease("linear")
    .attr("transform", "translate(" + this.x(0) + ")")
    .each("end", this.updateGraphData);
}

updateGraphData() {
  var newPoint = this.props.value;
  var newData = this.state.data.slice(1).concat([newPoint]);
  var newState = { data: newData, lineData: this.line(newData), source:"internal"};
  
  if (newPoint > this.state.maxVal) {
    newState.maxVal = newPoint;
  } else if (newPoint < this.state.minVal) {
    newState.minVal = newPoint;
  }
  this.setState(newState, () => {
    // the callback indicates the `setState` was complete
    // now we force a render
    this.forceUpdate();
  });

  requestAnimationFrame(this.updateGraph);
}
```

Since this component contains a mix of responsibilities in `render` (e.g. it renders the SVG stuff, but also some labels and stuff) I'd probably move the above pseudocode and the SVG-specific part of this into a dedicated sub-component, so that the `shouldComponentUpdate` hack affects the least code.